File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

void brute_force(vector<vector<int>>& arr, vector<int>& answer, int size,
                int y, int x) {
    
    //basecase
    if(size == 1){
        if(arr[y][x] == 1)
            answer[1]++;
        else
            answer[0]++;
        return;
    }
    
    
    // 전체 탐색
    bool isPossible = true;
    int prev = arr[y][x];
    for(int i = 0 ; i < size ; i++){
        for(int j = 0 ; j < size ; j++)
            if(arr[y + i][x + j] != prev){
                isPossible = false;
                break;
            }
        if(!isPossible)
            break;
    }
                
    if(isPossible){
        answer[prev]++;
        return;
    }
    else{
        brute_force(arr, answer, size / 2, y, x);
        brute_force(arr, answer, size / 2, y, x + (size / 2));
        brute_force(arr, answer, size / 2, y + (size / 2), x);
        brute_force(arr, answer, size / 2, (y + size /2), (x + size / 2));
        return;
    }
    
    return;
}


vector<int> solution(vector<vector<int>> arr) {
    vector<int> answer = {0, 0};
    brute_force(arr, answer, arr.size(), 0, 0);
    return answer;
}

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
import java.util.*;
import java.util.stream.*;

class Solution {
    
    int[] answer = new int[2];
        
    public int[] solution(int[][] arr) {
        
        compressArr(arr, 0, 0, arr.length);
        
        return answer;
    }
    
    
    public void compressArr(int[][] arr, int y, int x, int n){
        // 1칸이면 리턴
        if(n == 1){
            answer[arr[y][x]]++;
            return;
        }
                
        // 합 확인
        long sum = IntStream.range(0, n)
            .map(i -> IntStream.range(0, n)
                .map(j -> arr[y + i][x + j])
                .sum()
                )
            .sum();
        
        // 압축
        if(sum == 0){
            answer[0]++;
            return;
        }else if(sum == n * n){
            answer[1]++;
            return;
        }
        
        compressArr(arr, y, x, n / 2);
        compressArr(arr, y, x + n / 2, n / 2);
        compressArr(arr, y + n / 2, x, n / 2);
        compressArr(arr, y + n / 2, x + n / 2, n / 2);
        
        return;
    }
}